paraatm/simulation_method/vcas.py

```diff
@@ -0,0 +1,337 @@
+        self.track_file = cfg['data_file']
+        self.real = pd.read_csv(cfg['data_file'])
+
+        NatsEnvironment.start_jvm(nats_home=None)
+        self.NATS_SIMULATION_STATUS_PAUSE = NatsEnvironment.get_nats_constant('NATS_SIMULATION_STATUS_PAUSE')
+        self.NATS_SIMULATION_STATUS_ENDED = NatsEnvironment.get_nats_constant('NATS_SIMULATION_STATUS_ENDED')
+
+        natsStandalone = NatsEnvironment.get_nats_standalone()
+
+        self.simulationInterface = natsStandalone.getSimulationInterface()
+
+        self.entityInterface = natsStandalone.getEntityInterface()
+        self.controllerInterface = self.entityInterface.getControllerInterface()
+        self.pilotInterface = self.entityInterface.getPilotInterface()
+
+        self.environmentInterface = natsStandalone.getEnvironmentInterface()
+
+        self.equipmentInterface = natsStandalone.getEquipmentInterface()
+        self.aircraftInterface = self.equipmentInterface.getAircraftInterface()
+
+        if self.simulationInterface is None:
+            natsStandalone.stop()
+            raise RuntimeError("Can't get SimulationInterface")
+
+        self.simulationInterface.clear_trajectory()
+
+    def command_from_file(self):
+        """
+        extract commands related to altitude change
+
+        Returns
+        -------
+        Pandas DataFrame
+        """
+        text = pd.read_csv(self.cmd_file)
+        text = text.astype(object).replace(np.nan, 'None')
+        command_column = ['Timestamp', 'acID', 'action', 'altitude']
+        command_maintain = pd.DataFrame(columns=command_column)
+        for ind in range(len(text)):
+            line = text.loc[[ind]]
+            if 'maintain' in line['action1'].values[0] and line['new_alt'].values[0] != 'None':
+                timestamp = line['Timestamp'].values[0]
+                temp_c = [timestamp, line['acid'].values[0], line['action1'].values[0], line['new_alt'].values[0]]
+                command_maintain = command_maintain.append(pd.Series(temp_c, index=command_column), ignore_index=True)
+
+        cmd_maintain = command_maintain.sort_values(['Timestamp', 'acID'], ascending=True).reset_index(
+            drop=True)
+
+        return cmd_maintain
+
+    def get_start_time(self):
+        """
+        Extract simulation start time from input flight plan file into self variable
+        """
+        f = open(self.fp_file, 'r')
+        lines = f.readlines()
+        starttime = 999999999999
+        for x in lines:
+            if 'TRACK_TIME' in x:
+                ttemp = float(x[11:])
+                starttime = min(starttime, ttemp)
+
+        self.starttime = starttime
+
+        return
+
+    def simulation(self, *input_cmd, **kwargs):
+        """
+        Main function for simulation. NATS simulation based commands
+
+        Parameters
+        ----------
+        input_cmd : Pandas DataFrame
+            processed command in dataframe. If none, use command from input file
```

Comment:
Defining the argument as `*input_cmd` is not necessary and may not be doing what you think it is doing.  `NatsSimulationWrapper.__call__` passes additional arguments using `*args` and `**kwargs`, but they will appear to your function as regular arguments and/or keyword arguments.  You do not need to do anything special.  A more straightforward way to write this would be:
```python
def simulation(self, input_cmd=None):
        if input_cmd is None:
            cmd = self.command_from_file()
        else:
            cmd = input_cmd
```
When you define the argument as `*input_cmd`, the value `input_cmd` will hold a tuple of all non-keyword arguments.  So if you do `simulation(cmd)`, the value of `input_cmd` will actually be a tuple of length 1: `input_cmd = (cmd,)`.

Reply:
Thanks John, this is a brilliant idea! By assigning keyword for 'input_cmd' actually solved the problem I had when calling self() inside make_anti_model(). I will make the necessary changes and update this pull request. Thanks.

Reply:
Thanks, this is a brilliant idea! By assigning keyword for 'input_cmd' actually solved the problem I had when calling self() inside make_anti_model(). I will make the necessary changes and update this pull request. Thanks.